### DISCUSSION TITLE
how to access tcpstream in different tasks

### DISCUSSION THREAD

**Discussion by ActivePeter** [tokio-rs/tokio#4059] [Q&A] [answered]
I have a loop task to receive data and I saved the tcpstream rwlock to send with it in other tasks. But in the receive loop the socket is locked when calling the read so other tasks can not write.
![image](https://user-images.githubusercontent.com/8100696/130483627-5ac4fe71-fde2-4c2a-bc3f-0b64ab2c92fd.png)

**Comment by Darksonn** [comment, marked as answer]
Do not put the socket behind a mutex. For simple use-cases, you can use the `split` or `into_split` methods on the socket. For more complicated sharing, a common alternative is to use [the actor pattern][1].

[1]: https://ryhl.io/blog/actors-with-tokio/

**Reply by ActivePeter** [reply]
OK, thank you.
